### PR TITLE
BO: handle loading of extra product tabs

### PR DESCRIPTION
**Pull request by cdoublev**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.6.1.x
| Description?  | This allows modules developers to register an extra tab in admin product page (via hookDisplayAdminProductsExtra), without being forced to include their JS logic in their tab's template file, instead of using a separate JS file.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | Register a JS file via hookActionAdminControllerSetMedia to include in admin product page. It should be doing some basic stuff (eg. DOM content insert) inside the extra tab's content. Register this extra tab via hookDisplayAdminProductsExtra. Go to an admin product page. The JS file logic never happened.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/8139)

### CODE REVIEW

**Comment by cdoublev** [comment]
Hi,

I created this **~~very simple module~~** to demonstrate what this PR is solving.

The problem is not easy to expose briefly. I will try to add some context here by defining what the module should be doing, and by going through each development steps. **The module should append some content with Javascript inside an extra admin product tab.** It could have been providing autocompletion to an input text field, handling a file uploader, etc. as of other possible real use cases.

TL;DR
------

The only way to execute some Javascript interacting with an extra admin product tab is to include it in a `<script>` tag in the Smarty template of the extra admin product tab, and without waiting for any event to listen before firing its logic.

This PR allows extra admin product tabs to be handled by `productTabsManager` with the native admin product tabs, and to use external Javascript file to execute their needed interaction.

---

The module registers an external Javascript (JS) file to fullfill the previously defined goal. By default, the extra admin product tab's content is *"Needed content never appended."*. The external JS file responsibility is to replace it by the "needed content". If it works, it can be interpreted as "the needed JS feature has been loaded". 

After having installed the module, load an admin product page with your browser console open, and look at it + in `modules/cwtest/js/admin-product-tab.js` in your code editor or your browser sources panel, to see that this "needed content" has never been appended. Let's see why...

Each native admin product tabs (informations, associations, images, etc.) is declared as a JS function attached to a `product_tabs` global object defined in `js/admin/products.js`, which is included in document AFTER `modules/cwtest/js/admin-product-tab.js`. So we can't attach our own extra admin product tab on it by writing `product_tabs['ModuleMymodue'] = …` in `modules/cwtest/js/admin-product-tab.js`, because `product_tabs` will not be defined at this stage.

But what is `product_tabs` used for?

Any native admin product tabs that needs to execute some JS behavior when loaded, must have an `onReady` sub-function attached to itself. The reason of this is that all (native + extra) admin product tabs content are loaded async by `productTabsManager::display[Bulk]`, after window has been loaded, except the default/selected tab content which is loaded from the main request. These `onReady` sub-functions are fired when document is loaded or at the latest when the product tab is loaded. `productTabsManager::setTabs` is called (in `js/admin/products.js`) after each `product_tabs` has been declared, while document is loading. It registers each `product_tabs` it receives as its sole argument, in order to manage them. It could have been used to register our extra admin tab, but it's also defined AFTER `modules/cwtest/js/admin-product-tab.js` is included, and is not designed to register multiple tabs with several calls of it.

So the only option left is to not use an external file, but a `<script>` tag in the Smarty template of the extra admin product tab, and without waiting for any event to listen before firing its logic.

**Comment by xBorderie** [comment]
Wow, thanks for the details @cdoublev ! The team will soon have a look at your PR.

**Comment by cdoublev** [comment]
Thank you. Note that it's not really a bug fix but an improvement towards better coding practices. It's still possible to write `<script>initMyExtraAdminProductTab()</script>` and define `initMyExtraAdminProductTab` in an external file. The downside is global scope pollution and a risk of future collision between function names.

I (and my customers) was waiting for a review of this PR in order to push some commits (modules) in production, instead of using a local patch to be reverted later.

**Comment by cdoublev** [comment]
Hi,

I close this PR since LTS for 1.6 is coming relatively soon and I realized that using JS in template views (instead of in external files) is a practice used in many 1.6 native admin product tabs templates. I also forgot about `jQuery(productTabElement).ready()` which could be used to fire JS logic in time while keeping it in an external file without needing to fire it from view template. Anyway, small downsides.
